File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    # 각 장르별 총 재생 횟수를 저장하는 딕셔너리를 생성합니다.
    genre_play_dict = {}
    # 각 장르별 노래 정보를 저장하는 딕셔너리를 생성합니다.
    genre_index_dict = {}
    
    for i, (genre, play) in enumerate(zip(genres, plays)):
        # 각 장르별 총 재생 횟수를 계산합니다.
        genre_play_dict[genre] = genre_play_dict.get(genre, 0) + play
        # 각 장르별 노래 정보를 저장합니다.
        if genre not in genre_index_dict:
            genre_index_dict[genre] = []
        genre_index_dict[genre].append((i, play))
    
    # 총 재생 횟수가 많은 장르부터 처리하기 위해 정렬합니다.
    sorted_genres = sorted(genre_play_dict.keys(), key=lambda x: genre_play_dict[x], reverse=True)
    
    answer = []
    for genre in sorted_genres:
        # 각 장르 내에서 많이 재생된 노래부터 처리하기 위해 정렬합니다.
        sorted_genre_songs = sorted(genre_index_dict[genre], key=lambda x: x[1], reverse=True)
        # 최대 두 개의 노래만 선택합니다.
        for i, (index, _) in enumerate(sorted_genre_songs):
            if i >= 2:
                break
            answer.append(index)
    
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {
    class Song implements Comparable<Song> {
        int id;
        int play;
        
        Song(int id, int play) {
            this.id = id;
            this.play = play;
        }
        
        @Override
        public int compareTo(Song o) {
            if (this.play == o.play) {
                return this.id - o.id;
            }
            return o.play - this.play;
        }
    }
    public int[] solution(String[] genres, int[] plays) {
        Map<String, Integer> totalMap = new HashMap<>();
        Map<String, List<Song>> songMap = new HashMap<>();
        List<Integer> result = new ArrayList<>();
        for (int i = 0; i < genres.length; i++) {
            String genre = genres[i];
            
            totalMap.put(genre, totalMap.getOrDefault(genre, 0 ) + plays[i]);
            songMap.putIfAbsent(genre, new ArrayList<>());
            songMap.get(genre).add(new Song(i, plays[i]));
        }
        List<String> genreList = new ArrayList<>(totalMap.keySet());
        
        genreList.sort((a,b) -> totalMap.get(b) - totalMap.get(a));
        
        for (String genre : genreList) {
            List<Song> songs = songMap.get(genre);
            Collections.sort(songs);
            
            for (int i = 0; i < Math.min(2, songs.size()); i++){
                result.add(songs.get(i).id);
            }
        }
        return result.stream().mapToInt(i->i).toArray();
    }
}